samza-azure/src/main/java/org/apache/samza/system/azureblob/producer/AzureBlobSystemProducer.java

```diff
@@ -0,0 +1,505 @@
+          new BlobServiceClientBuilder()
+          .httpLogOptions(httpLogOptions)
+          .endpoint(endpoint)
+          .credential(credential)
+          .httpClient(httpClient)
+          .buildAsyncClient();
+
+
+      SkuName accountType = storageClient.getAccountInfo().block().getSkuName();
+      long flushThresholdSize = config.getMaxFlushThresholdSize(systemName);
+      boolean isPremiumAccount = SkuName.PREMIUM_LRS == accountType;
+      if (isPremiumAccount && flushThresholdSize > PREMIUM_MAX_BLOCK_SIZE) { // 100 MB
+        throw new SystemProducerException("Azure storage account with name: " + accountName
+            + " is a premium account and can only handle upto 100MB threshold size. Given flush threshold size is "
+            + flushThresholdSize);
+      } else if (!isPremiumAccount && flushThresholdSize > STANDARD_MAX_BLOCK_SIZE) { // STANDARD account
+        throw new SystemProducerException("Azure storage account with name: " + accountName
+            + " is a standard account and can only handle upto 4MB threshold size. Given flush threshold size is "
+            + flushThresholdSize);
+      }
+
+      containerAsyncClient = storageClient.getBlobContainerAsyncClient(systemName);
+
+      // Only way to check if container exists or not is by creating it and look for failure/success.
+      createContainerIfNotExists(containerAsyncClient);
+    } catch (Exception e) {
+      metrics.updateAzureContainerMetrics();
+      throw new SystemProducerException("Failed to set up Azure container for SystemName: " + systemName, e);
+    }
+  }
+
+  /**
+   * // find the writer in the writerMap else create one
+   * @param source for which to find/create the writer
+   * @param messageEnvelope to fetch the schema from if writer needs to be created
+   * @return an AzureBlobWriter object
+   */
+  @VisibleForTesting
+  AzureBlobWriter getOrCreateWriter(String source, OutgoingMessageEnvelope messageEnvelope) {
+    String writerMapKey;
+    String blobURLPrefix;
+    String partitionKey = getPartitionKey(messageEnvelope);
+    // using most significant bits in UUID (8 digits) to avoid collision in blob names
+    if (partitionKey == null) {
+      writerMapKey = messageEnvelope.getSystemStream().getStream();
+      blobURLPrefix = String.format(BLOB_NAME_PREFIX, messageEnvelope.getSystemStream().getStream());
+    } else {
+      writerMapKey = messageEnvelope.getSystemStream().getStream() + "/" + partitionKey;
+      blobURLPrefix = String.format(BLOB_NAME_PARTITION_PREFIX, messageEnvelope.getSystemStream().getStream(), partitionKey);
+    }
+    Map<String, AzureBlobWriter> sourceWriterMap = writerMap.get(source);
+    if (sourceWriterMap == null) {
+      throw new SystemProducerException("Attempting to send to source: " + source + " but it is not registered");
+    }
+    AzureBlobWriter writer = sourceWriterMap.get(writerMapKey);
+    if (writer == null) {
+      synchronized (sourceWriterCreationLockMap.get(source)) {
```

Comment:
Should these two lines be switched? Otherwise, two different threads could still create a new writer.

Reply:
the writer creation is wrapped inside another check if writer == null at line 410 and hence 2 thread will not create the writer.
On the other hand if we did switch the statements, then every thread every time will be forced to acquire the lock even if the writer is not null

here is the problematic scenario and why it wont lead to two writer creation 
at t = 0, two threads T1 and T2 both hit line 407 and both see writer = null
at t = 1, T1 and T2 both hit line 408 and try to get lock
at t = 2, T1 gets the lock and T2 enters wait
at t = 3, T1 hits line 409 and gets the writer from the map again and it is still null
at t =4 , T1 hits line 410 and checks if writer is null
at t = 5, T1 creates writer in line 411,412 and 413 
at t = 6, T1 adds the writer to the map
at t= 7, T1 exits critical section and gives up lock and  T2 gets it. 
at t = 8, T2 hits 409 and gets writer from map
at t = 9, T2 checks if writer is null and it is not
at t = 10, T2 exits the critical section and returns the writer

Reply:
Aha, I missed that you check for `writer == null` again inside the synchronized block.